common/flatpak-transaction.c

```diff
@@ -4081,6 +4114,325 @@ flatpak_transaction_normalize_ops (FlatpakTransaction *self)
+              g_ptr_array_add (run_after_ops, op);
+            }
+          else
+            g_hash_table_add (newly_used_runtimes, g_strdup (runtime_ref));
+
+          if (!priv->disable_related)
+            {
+              for (i = 0; runtime_related && i < runtime_related->len; i++)
+                {
+                  FlatpakRelated *rel = g_ptr_array_index (runtime_related, i);
+
+                  if (op_type == FLATPAK_TRANSACTION_OPERATION_UNINSTALL)
+                    {
+                      if (rel->delete)
+                        g_hash_table_replace (maybe_unused_runtimes, g_strdup (rel->ref), g_strdup (runtime_remote));
+                    }
+                  else
+                    g_hash_table_add (newly_used_runtimes, g_strdup (rel->ref));
+                }
+            }
+        }
+
+      if (op_type == FLATPAK_TRANSACTION_OPERATION_UPDATE &&
+          dir_ref_is_installed (priv->dir, op->ref, NULL, &deploy_data))
+        {
+          g_autofree char *full_previous_runtime = NULL;
+          g_autofree char *prev_runtime_remote = NULL;
+          g_autoptr(GPtrArray) runtime_related = NULL;
+          g_autoptr(GPtrArray) ref_related = NULL;
+
+          g_ptr_array_add (run_after_ops, op);
+
+          /* The related refs of the op might now be unused */
+          if (!priv->disable_related)
+            {
+              ref_related = find_related_from_deploy (self, op->ref, NULL);
+              for (i = 0; ref_related && i < ref_related->len; i++)
+                {
+                  FlatpakRelated *rel = g_ptr_array_index (ref_related, i);
+                  if (rel->delete)
+                    g_hash_table_replace (maybe_unused_runtimes, g_strdup (rel->ref), g_strdup (op->remote));
+                }
+            }
+
+          /* If the op is changing to a different runtime, the previous one and
+           * its related refs might now be unused */
+          const char *previous_runtime = flatpak_deploy_data_get_runtime (deploy_data);
+
+          if (previous_runtime == NULL || *previous_runtime == '\0')
+            continue;
+
+          full_previous_runtime = g_strconcat ("runtime/", previous_runtime, NULL);
+          if (g_strcmp0 (full_previous_runtime, runtime_ref) == 0)
+            continue;
+
+          runtime_related = find_related_from_deploy (self, full_previous_runtime, &prev_runtime_remote);
+
+          g_hash_table_replace (maybe_unused_runtimes,
+                                g_strdup (full_previous_runtime),
+                                g_strdup (prev_runtime_remote));
+
+          if (priv->disable_related)
+            continue;
+
+          for (i = 0; runtime_related && i < runtime_related->len; i++)
+            {
+              FlatpakRelated *rel = g_ptr_array_index (runtime_related, i);
+              if (rel->delete)
+                g_hash_table_replace (maybe_unused_runtimes, g_strdup (rel->ref), g_strdup (prev_runtime_remote));
+            }
+        }
+    }
```

Comment:
I think this is quite complex and it is not obvious we're not missing anything else. I wonder if we could instead replace it with calling find_unused() twice. Once for whatever is currently installed and then again with the overrides and excludes for the operations. If with the overrides any ref is show as unused, but it wasn't without, then we can uninstall it.

That seems easier to reason about and more flexible to changes in how dependencies work.